riverschematest/riverschematest.go

```diff
@@ -0,0 +1,377 @@
+// are truncated to leave a clean state for the next test.
+//
+// Use of a schema doesn't happen in River clients automatically. TestSchema
+// returns the name of a schema to use. This should be set in River clients,
+// used as a parameter for testfactory functions, and anywhere else where
+// database operations are issued.
+//
+// Despite reasonably fast migrations and efficient reuse, test schemas still
+// have more overhead than a test transaction, so prefer the use of
+// riversharedtest.TestTx where a full database isn't needed (TestSchema is
+// commonly needed where tests rely on database notifications).
+//
+// tb is an interface that tolerates not having a cleanup hook so constructs
+// like testutil.PanicTB can be used. If Cleanup isn't available, schemas aren't
+// checked back in for reuse.
+//
+// Where Cleanup is available, the function also performs a post-flight check
+// that makes sure all tables in TruncateTables are empty. This helps detect
+// problems where test cases accidentally inject data into the default schema
+// rather than the one returned by this function.
+func TestSchema[TTx any](ctx context.Context, tb testutil.TestingTB, driver riverdriver.Driver[TTx], opts *TestSchemaOpts) string {
+	tb.Helper()
+
+	require.NotNil(tb, driver, "driver should not be nil")
+
+	if opts == nil {
+		opts = &TestSchemaOpts{}
+	}
+
+	// An initial pass to calculate a friendly package name that'll be used to
+	// prefix this package's schemas so that it won't clash with packages
+	// running their own tests in parallel. Generated name is like `river` or
+	// `jobcompleter` or `riverpro`.
+	genSchemaBase.Do(func() {
+		var (
+			programCounterAddr, _, _, _ = runtime.Caller(4)                            // skip `TestSchema.func1` (closure) + `sync.(*Once).doSlow` + `sync.(*Once).Do` + `TestSchema` and end up at `TestSchema`'s caller
+			funcName                    = runtime.FuncForPC(programCounterAddr).Name() // like: github.com/riverqueue/river.Test_Client.func1
+		)
+
+		packageName = packageFromFunc(funcName) // like: `river` (or `jobcompleter`, or `riverpro`)
+
+		// Check to make sure we're skipping the right number of frames above.
+		// If the location of `runtime.Caller` is changed at all (a single new
+		// function is added to the stack), the reported package will be
+		// completely wrong, so we try to take precautions about it.
+		if packageName == "riverschematest" && !opts.skipPackageNameCheck {
+			panic("package name should not resolve to riverschematest")
+		}
+
+		// Notification topics are prefixed with schemas. The max Postgres
+		// length of topics is 63, and "river_leadership" is the longest topic
+		// name. If the package suffix is longer than the max that could fit
+		// into 63 when combined with a schema name and leadership, trim it down
+		// a bit. The only package where this is needed as I write this is
+		// `riverencrypt_test`. If this happens in too many places we may want
```

Comment:
Do we need to validate the length of the schema param in the river config to avoid issues with this limit?

Reply:
Yeah good call. I just put that in.